sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/controller/container/ThroughputContainerController.java

```diff
@@ -0,0 +1,321 @@
+            .single()
+            .flatMap(offerFeedResponse -> {
+                if (offerFeedResponse.getResults().isEmpty()) {
+                    return Mono.error(
+                        BridgeInternal.createCosmosException(NO_OFFER_EXCEPTION_STATUS_CODE, "No offers found for the resource " + resourceId));
+                }
+
+                return this.client.readOffer(offerFeedResponse.getResults().get(0).getSelfLink()).single();
+            })
+            .onErrorResume(throwable -> {
+                System.out.println(throwable.getCause());
+                return Mono.error(throwable);
+            })
+            .map(ModelBridgeInternal::createThroughputRespose);
+    }
+
+    private void setDefaultGroupController() {
+        List<ThroughputGroupControllerBase> defaultGroupControllers =
+            this.groupControllers.values().stream().filter(ThroughputGroupControllerBase::isUseByDefault).collect(Collectors.toList());
+
+        if (defaultGroupControllers.size() > 1) {
+            // We should never reach here since we do config validation to make sure only one default group configured
+            // when enable throughput control
+
+            throw new IllegalArgumentException("There should only be one default throughput control group");
+        }
+        if (defaultGroupControllers.size() == 1) {
+            this.defaultGroupController = defaultGroupControllers.get(0);
+        }
+    }
+
+    private void updateMaxContainerThroughput(ThroughputResponse throughputResponse) {
+        checkNotNull(throughputResponse, "Throughput response can not be null");
+
+        ThroughputProperties throughputProperties = throughputResponse.getProperties();
+        this.maxContainerThroughput.set(
+            Math.max(throughputProperties.getAutoscaleMaxThroughput(), throughputProperties.getManualThroughput()));
+    }
+
+    private boolean isOfferNotConfiguredException(Throwable throwable) {
+        checkNotNull(throwable, "Throwable should not be null");
+
+        CosmosException cosmosException = Utils.as(throwable, CosmosException.class);
+        // the exception here should match what returned from method resolveThroughputByResourceId
+        return cosmosException != null
+            && cosmosException.getStatusCode() == NO_OFFER_EXCEPTION_STATUS_CODE
+            && cosmosException.getSubStatusCode() == NO_OFFER_EXCEPTION_SUB_STATUS_CODE;
+    }
+
+    @Override
+    public <T> Mono<T> processRequest(RxDocumentServiceRequest request, Mono<T> originalRequestMono) {
+        checkNotNull(request, "Request can not be null");
+        checkNotNull(originalRequestMono, "Original request mono can not be null");
+
+        return Mono.just(request)
+            .flatMap(request1 -> {
+                if (request1.getThroughputControlGroupName() == null) {
+                    return Mono.just(new Utils.ValueHolder<>(this.defaultGroupController));
```

Comment:
We only use `ValueHolder` in case where we think that value inside a Flux or Mono can be null, since reactor doesn't support null values in Flux or Mono.  If that is not the case here, then we don't need to use `ValueHolder` here.

Reply:
yup, since defaultGroupController can be null, so intend to use valueHolder for null scenario.